### ISSUE TITLE
optimisations for prefetching chunks

### ISSUE THREAD

**Issue by danielroe** [nuxt/framework#6638]
Currently dynamic imports include things like routes, error components, and any global components (so if using with `@nuxt/content`). I am _very_ concerned to decrease HTML and number of prefetched resources but if we simply disable prefetching, sites will seem much slower.

From my point of view, before removing current prefetch behaviour, we need the ability to make some finer-grained decisions.

A set of suggestions:

- [ ] `prefetch` support via NuxtLink (https://github.com/nuxt/framework/pull/4329)
- [ ] programmatic prefetching of routes: https://github.com/nuxt/framework/issues/3649
- [ ] remove page chunks from dynamic imports in manifest, or smartly add 'sibling' pages
- [ ] programmatic prefetching of global components
- [ ] (optionally) add a plugin to update manifest so that the components that _might_ use global components are the ones that have the dynamic import registered
- [ ] add hook for user to manipulate client manifest
- [ ] (possibly) add ability for user to customise `vue-bundle-renderer` options - see https://github.com/nuxt/framework/pull/6210#discussion_r937950236.

**Comment by kosmeln** [comment]
Hello @danielroe .
I'm not sure if this is related to this improvement or I should create a separate issue. Please let me know.
I've created a [test project](https://codesandbox.io/s/nuxt3-playground-rc-8-8xvt77?file=/pages/red-page.vue) to visualise how pre-fetching stylesheets may affect the way page looks.
There are 2 layouts - red and blue (under /layouts). If I open red-page component (https://8xvt77.sse.codesandbox.io/red-page), blue layout - that is being pre-fetched - affects the way page looks - and h2 that should be black displays in blue instead.
The example is oversimplified but we have a real use case of this being a problem for our project.

**Comment by pi0** [comment]
@kosmeln Please ensure to use scoped styles (`<style scoped>`) or selector prefixes per parent/layout. You can use a special CSS class in layouts that enable them otherwise, loading a layout can make unpredictable behavior. See this fork: https://codesandbox.io/s/nuxt3-playground-rc-8-forked-6l6uot?file=/layouts/blue.vue:122-134

**Comment by pi0** [comment]
@danielroe As we discussed, we can move both of first items in parallel.

Build-time and manifest based page prefetching is probably something we don't want to do in Nuxt 3 since was always tricky in Nuxt 2 when number of pages increases. Only reliable way to predict next pages is runtime rendering. Both Server-Side and Client-Side of `NuxtLink` can smartly add prefetches.

**Comment by kosmeln** [comment]
> @kosmeln Please ensure to use scoped styles (`<style scoped>`) or selector prefixes per parent/layout. You can use a special CSS class in layouts that enable them otherwise, loading a layout can make unpredictable behavior. See this fork: https://codesandbox.io/s/nuxt3-playground-rc-8-forked-6l6uot?file=/layouts/blue.vue:122-134

Thanks for the quick response @pi. 

I do aware of this solution and this is what I currently use as a workaround for my project.
The thing is - in my current project we use single NUXT app for multiple sites. We have 1 layout per 1 site.
Each site uses same vue components as the underlaying logic is mostly the same. We use the layout split purely for css tweaks.
We use global styles that only contains the css difference for a current site in addition to loading same common styles for all sites. So the global styles are used as an override for some common css rules.
FYI: When we used NUXT2 this worked perfectly with setting
```
splitChunks: {
    layouts: true
}
```
Speaking of other use cases, there is an example posted by @engvie of using similar approach when splitting between site and admin pages - see [this thread](https://github.com/nuxt/framework/issues/4475#event-7141976773). 

The workaround that we use (same as you shared) resolves the issue, but the problem is - page pre-fetches all layouts (for my case a dozen of them) and they will never be needed in per-site approach - so this is just the waste of resources.

I really hope that there is going to be any mechanisms developed to handle this use-case either in the scope of this ticket or further on. Until then - any other suggestions that you might have are very welcomed.

**Comment by pi0** [comment]
Thanks for explaining your use case @kosmeln. Prefetching is not causing any issue but simply _reveals_ a side-effect issue of not using scoped styles in layouts. Using style prefixes is not a workaround but a proper way to write global styles that ensure they don't conflict.

I would be more than happy to continue this conversation to investigate different solutions we might provide for this. If you would, please open a [discussion](https://github.com/nuxt/framework/discussions) and elaborate more about site builder use-case that ends up with dozens of layouts. 

BTW in Nuxt 3, we have a _much better solution_ for theming and shared code for multi sites depending on same code base but slight changes. It is called [extends](https://github.com/nuxt/framework/issues/3222).

**Comment by jshimkoski** [comment]
I believe this to be related: https://github.com/nuxt/framework/discussions/9463

As you can see in that discussion, simply removing prefetch tags from my app increases its Lighthouse score by 20%. And I can tell you the app feels and loads faster as well in use.

If developers could gain access to `shouldPrefetch` and `shouldPreload`, that would really help in ginormous applications like mine where prefetch actually leads to performance loss instead of gain.
